### PR TITLE
hyperparameter-optimization-Maxie-M.py

### PR DESCRIPTION
**Pull request by Maxie-M**

### CODE REVIEW

**Comment by larskotthoff** [comment]
As I've said multiple times before, please *do not* open new pull requests. You can change the already existing pull requests.

**Review by larskotthoff — CHANGES_REQUESTED** [review]
Can you say why performance is worse after tuning and what you could do to improve it?

**Comment by Maxie-M** [comment]
Please let me know if you would like more detail, I tried answering everything you wanted to the best of my abilities. I'm sorry for the misunderstandings.

**Review by larskotthoff — CHANGES_REQUESTED** [review]
- You're only running 10 iterations -- if you increase that, do you get better results?
- How would a CV help you to get better results?

**Comment by Maxie-M** [comment]
I'm really sorry but I'm very confused. Are you asking me to change my code, I thought the only issue was with the paper because you never once requested anything to be changed in my code so it's been the same exact code for the last few turn-ins. If you aren't asking me to change my code, and again requesting more changes on my paper, how am I supposed to give you an accurate answer if more iterations will give me better results?

I to my knowledge fulfilled all of your change requests to the best of my capabilities. For example this new request seems to. be an add on to the two previous requests by you:

1. "Sorry for being unclear -- your description should contain enough detail to reproduce what you did. Please describe the ranges of the hyperparameter values that you tuned, i.e. what was the minimum and maximum number of estimators? Similarly, how many CV folds did you use in RandomSearchCV?

And why does performance get worse after optimization?"

2. "Can you say why performance is worse after tuning and what you could do to improve it?" 

Which I thought I successfully fulfilled. Which these changes can be seen in these following parts of my paper which is is the most recent upload: 

1. " Two machine learning algorithms were chosen for the model training and performing the hyperparameter optimization. The algorithms selected were the Decision Tree Classifier and Random Forest Classifier. For this a search space was created for hyperparameters, this included: 

Decision Tree Classifier:
Maximum Depth:                 Ranging from None to 15
Maximum Features:            ‘auto’, ‘sqrt’, ‘log2’, and None
Minimum Samples Split:     Varying from 2 to 10
Minimum Samples Leaf:     Ranging from 1 to 4

Random Forest Classifier:
Maximum Depth:                  Ranging from None to 15
Maximum Features:             ‘auto’, ‘sqrt’, ‘log2’, and None
Number of Estimators:         Ranging from 50 to 150
Minimum Samples Split:      Varying from 2 to 10
Minimum Samples Leaf:       Ranging from 1 to 4

For both the Decision Tree Classifier and the Random Forest Classifier. The Random Forest Classifier also will be considering n_estimators, meaning it will consider the number of trees in the forest. On top of this RandomSearchCV has been utilized to perform nested resampling for hyperparameter optimization. The  way this works is like double-checking where data is split into multiple sets of training and testing and further. Now for each of these splits it will be further split for validation in finding the best parameters. Helping assess models accuracy without overfitting. This was chosen due to its efficiency in the exploration of the hyperparameter space. For this exercise there was a total of 10 iterations and 5-fold cross-validation in order to conduct this search. Now, to be able to visualize this it was chosen to incorporate ROC curves. Plotting the ROC curves will utilize the Area Under the Curve (AUC) and visuale the performance of the hyperparameter tuning and each model [fig. 2]. Then a heatmap will be used in visualizing the best configuration of hyperparameters for each model."

2. "While there were obvious reductions when it comes to accuracy after optimizing both the decision tree classifier and random forest this plays into an important aspect of hyperparameter optimization. This is: improvement in performance is not guaranteed in all cases. Some factors that may have contributed to this phenomenon are:
      Complexity/Quality of the data 
      May lead to overfitting 
      Insufficient search space exploration 
The quality of the data could’ve been what impacted the model performance. I personally did not remove the outlier within the dataset which could have affected the model’s ability to generalize. Another point to recognize hyperparameter tuning may lead to overfitting. This happens if the models have become too specialized to the training set. This would definitely decrease the performance on the unseen data. Lastly the range that was used for what hyperparameters explored during the optimization could’ve been too narrow/limited. The ways this could be improved on consist of the following: 
       Feature Engineering 
       Increasing Search Space
       Cross-Validation
The exploration of feature engineering techniques may help improve performance. This is due to the creation of informative features or the reduction of dimensionality of the dataset. Doing this could potentially improve on the model’s ability to generalize.Next would be the recommendation of increasing the search space. Expanding the range of what hyperparameters would be explored during the optimization. This could potentially give a more comprehensive search when it comes to the configurations. Lastly the implementation of cross-validation could potentially improve the performance. Implementing cross-validation techniques could ensure that the model is not overly dependent on one subset of the data. Overall despite these obvious decreases within the accuracy, the tuned models still provide and maintain competitive performance. This is with the indications that the optimization of the hyperparameters process provided refined model configurations with a minor sacrifice of accuracy. " 

I believe this is the clearest I can be with all of your request without having to create a second code to actually apply these suggestions I have made to see if it would improve the performance and/or results. 

Also is this the last thing I need to do to get this exercise approved? As you stated on multiple occasions you don't want to do back and forth and will stop even grading at a certain point. On top of this I only get three turn-ins a week which I would preferably reserve to be able to turn in the next two exercises once those are completed. So if possible if you have more things you would like me to do could you please provide my corrections all at once so I am no longer wasting your time having to do one correction at a time.

**Comment by larskotthoff** [comment]
Don't worry about the limit on turn ins etc, the main thing is to get the result right.

The change to run more iterations should be trivial (one number in line 80 if I see this correctly), and as we've discussed in class several times, it is important to show that you've ensured to have a good chance of improving over the default configurations. At the moment you do have a lot of explanation, but all of this is speculation as far as I can see, and testing some of your hypotheses there (e.g. exploring more of the search space) is very easy.

**Comment by Maxie-M** [comment]
I see, I’m sorry for all the misunderstanding, I went back into my code to try more iterations which actually made the performance worse! And then I realized I was using the default accuracy for my hyperparameter optimization, instead of what I should have been using, which was the roc_auc scoring. Once I fixed that I also decided to add StratifiedKFold to ensure there was no set being favored. With these two changes I seem to have fixed my results, now showing that the hyperparameter optimization did indeed improve performance of models.

**Review by larskotthoff — APPROVED** [review]
